File: flyweight/Flyweight.ts
class BulletType {
    speed: number;
    damage: number;
    name: string;

    constructor(name: string, speed: number, damage: number) {
        this.name = name;
        this.speed = speed;
        this.damage = damage;
    }
}

class Bullet {
    x: number;
    y: number;
    type: BulletType;

    constructor(x: number, y: number, type: BulletType) {
        this.x = x;
        this.y = y;
        this.type = type;
    }

    describe() {
        console.log(`Bullet ${this.type.name}`);
        console.log(`Speed ${this.type.speed}`);
        console.log(`Damage ${this.type.damage}`);
        console.log(`@ (${this.x}, ${this.y})`);
    }
}

class BulletFactory {
    bulletTypes: Map<string, BulletType> = new Map();

    getBullet(name: string, x: number, y: number) {
        let type = this.bulletTypes.get(name);
        if (!type) {
            type = new BulletType(name, 10, 10);
            this.bulletTypes.set(name, type);
        }
        return new Bullet(x, y, type);
    }
}

function testFlyweight() {
    const factory = new BulletFactory();
    const bullet1 = factory.getBullet("AK47", 0, 0);
    const bullet2 = factory.getBullet("AK47", 0, 10);
    bullet1.describe();
    bullet2.describe();

    const bullet3 = factory.getBullet("M4A1", 0, 10);
    const bullet4 = factory.getBullet("M4A1", 0, 20);

    bullet3.describe();
    bullet4.describe();
}

testFlyweight();

File: flyweight/Flyweight.py
class BulletType:
    name: str
    speed: int
    damage: int

    def __init__(self, name, speed, damage):
        self.name = name
        self.speed = speed
        self.damage = damage


class Bullet:
    x: int
    y: int
    type: BulletType

    def __init__(self, x, y, type):
        self.x = x
        self.y = y
        self.type = type

    def describe(self):
        print("Bullet " + self.type.name)
        print("Speed " + str(self.type.speed))
        print("Damage " + str(self.type.damage))
        print("@ (" + str(self.x) + ", " + str(self.y) + ")")


class BulletFactory:
    bulletTypes = {}

    def getBullet(self, name, x, y):
        type = self.bulletTypes.get(name)
        if type == None:
            type = BulletType(name, 10, 10)
            self.bulletTypes[name] = type
        return Bullet(x, y, type)


def testFlyweight():
    factory = BulletFactory()
    bullet1 = factory.getBullet("AK47", 0, 0)
    bullet2 = factory.getBullet("AK47", 0, 10)
    bullet1.describe()
    bullet2.describe()

    bullet3 = factory.getBullet("M4A1", 0, 10)
    bullet4 = factory.getBullet("M4A1", 0, 20)

    bullet3.describe()
    bullet4.describe()


if __name__ == "__main__":
    testFlyweight()
